hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java

```diff
@@ -362,11 +508,375 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
         abortWriter(mobFileWriter);
       }
     }
+    // Commit or abort generational writers
+    if (mobWriters != null) {
+      for (StoreFileWriter w: mobWriters.getOutputWriters()) {
+        Long mobs = mobWriters.getMobCountForOutputWriter(w);
+        if (mobs != null && mobs > 0) {
+          mobRefSet.get().add(w.getPath().getName());
+          w.appendMetadata(fd.maxSeqId, major, mobs);
+          w.close();
+          mobStore.commitFile(w.getPath(), path);
+        } else {
+          abortWriter(w);
+        }
+      }
+    }
     mobStore.updateCellsCountCompactedFromMob(cellsCountCompactedFromMob);
     mobStore.updateCellsCountCompactedToMob(cellsCountCompactedToMob);
     mobStore.updateCellsSizeCompactedFromMob(cellsSizeCompactedFromMob);
     mobStore.updateCellsSizeCompactedToMob(cellsSizeCompactedToMob);
     progress.complete();
     return true;
   }
+
+  protected static String createKey(TableName tableName, String encodedName,
+      String columnFamilyName) {
+    return tableName.getNameAsString()+ "_" + encodedName + "_"+ columnFamilyName;
+  }
+
+  @Override
+  protected List<Path> commitWriter(StoreFileWriter writer, FileDetails fd,
+      CompactionRequestImpl request) throws IOException {
+    List<Path> newFiles = Lists.newArrayList(writer.getPath());
+    writer.appendMetadata(fd.maxSeqId, request.isAllFiles(), request.getFiles());
+    // Append MOB references
+    Set<String> refSet = mobRefSet.get();
+    writer.appendMobMetadata(refSet);
+    writer.close();
+    return newFiles;
+  }
+
+  private List<Path> getReferencedMobFiles(Collection<HStoreFile> storeFiles) {
+    Path mobDir = MobUtils.getMobFamilyPath(conf, store.getTableName(), store.getColumnFamilyName());
+    Set<String> mobSet = new HashSet<String>();
+    for (HStoreFile sf: storeFiles) {
+      byte[] value = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+      if (value != null) {
+        String s = new String(value);
```

Comment:
this should be `Bytes.toString(byte[])` instead of relying on the platform encoding.